lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -396,21 +379,326 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+      if (innerLoop) {
+        OpBuilder innerLoopRegion(loopNest.loops.back().getRegion());
+        SmallVector<Value> innerIndices;
+        int i = 0;
+        for (auto [affineMap, iterator] : iteratorMaps) {
+          // Calculate how far a single step in the inner dim is.
+          SmallVector<std::int64_t> innerOffset =
+              affineMap.compose(innerIndexStep);
+          assert(innerOffset.size() == 1);
+          innerIndices.push_back(innerLoopRegion.create<arith::AddIOp>(
+              output.getLoc(), forOp.getRegionIterArg(i),
+              i32(innerOffset[0], innerLoopRegion)));
+          ++i;
+        }
+        innerLoopRegion.create<scf::YieldOp>(output.getLoc(), innerIndices);
+      }
+
+      // Backpedal and adjust the iterator offset for the current dimension.
+      if (dim > 0) {
+        SmallVector<Value> outerIndices;
+        SmallVector<std::int64_t> outerIndexStep(outputAffineMap.getNumDims(),
+                                                 0);
+        outerIndexStep[dim - 1] = 1;
+        int i = 0;
+        for (auto [affineMap, iterator] : iteratorMaps) {
+          // Calculate how far a single step in the inner dim is.
+          SmallVector<std::int64_t> innerOffset =
+              affineMap.compose(innerIndexStep);
+          assert(innerOffset.size() == 1);
+          // Calculate how far a single step in the outer dim is.
+          SmallVector<std::int64_t> outerOffset =
+              affineMap.compose(outerIndexStep);
+          assert(outerOffset.size() == 1);
+          // Multiply by the number of steps that the inner loop took.
+          // FIXME: test this for higher dims
+          std::int64_t offset =
+              outerOffset[0] - innerOffset[0] * outputShape[dim];
+          outerIndices.push_back(regionBuilder.create<arith::AddIOp>(
+              output.getLoc(), forOp.getResult(i), i32(offset, regionBuilder)));
+          ++i;
+        }
+        regionBuilder.create<scf::YieldOp>(output.getLoc(), outerIndices);
+      }
+
+      loopNest.loopRegions.push_back(&loopNest.loops.back().getRegion());
+      iteratorsNest.emplace_back(forOp.getRegionIterArgs());
+    }
+
+    return loopNest;
+  }
+
+  // Convert arith and math dialect operations into ttkernel init tile
+  // operations. HLK requires the FPU to be initialized before any tile ops get
+  // executed.  We separate the init tile operation from the actual tile
+  // operation so that we can hoist the init tile operation outside of the loop
+  // nest.
+  void convertComputeInitOp(Operation &op, ArrayRef<BlockArgument> cbOperands,
+                            std::int64_t numDpsInputs,
+                            OpBuilder &builder) const {
+    SmallVector<std::int64_t> operandIndices;
+    for (OpOperand &operand : op.getOpOperands()) {
+      operandIndices.push_back(operand.getOperandNumber());
+    }
+    if (mlir::isa<arith::AddFOp>(op)) {
+      builder.create<ttkernel::BinaryOpInitCommonOp>(
```

Comment:
You can extract `BinaryOpInitCommonOp` outside of if condition.

Reply:
In the future there might be `else if` cases that are unary functions